src/machine/parsed_results.rs

```diff
@@ -141,6 +79,206 @@ pub enum Value {
     Var(String),
 }
 
+impl Value {
+    pub fn conjunction(conj: &[Value]) -> Value {
+        if conj.is_empty() {
+            panic!("can't create empty conjunction");
+        } else if conj.len() == 1 {
+            return conj[0].clone();
+        }
+
+        let mut conj_rev = conj.iter().rev();
+
+        let last = conj_rev.next().unwrap().clone();
+        let beflast = conj_rev.next().unwrap().clone();
+        let mut conj_term = Value::Structure(",".into(), vec![beflast, last]);
+
+        for val in conj_rev {
+            conj_term = Value::Structure(",".into(), vec![val.clone(), conj_term]);
+        }
+
+        conj_term
+    }
+
+    pub fn disjunction(disj: &[Value]) -> Value {
+        if disj.is_empty() {
+            panic!("can't create empty disjunction");
+        } else if disj.len() == 1 {
+            return disj[0].clone();
+        }
+
+        let mut disj_rev = disj.iter().rev();
+
+        let last = disj_rev.next().unwrap().clone();
+        let beflast = disj_rev.next().unwrap().clone();
+        let mut disj_term = Value::Structure(";".into(), vec![beflast, last]);
+
+        for val in disj_rev {
+            disj_term = Value::Structure(";".into(), vec![val.clone(), disj_term]);
+        }
+
+        disj_term
+    }
+}
+
+impl From<&QueryResolutionLine> for Value {
+    fn from(qrl: &QueryResolutionLine) -> Self {
+        match qrl {
+            QueryResolutionLine::True => Value::Atom("true".into()),
+            QueryResolutionLine::False => Value::Atom("false".into()),
+            QueryResolutionLine::Match(m) => Value::conjunction(
+                &m.bindings
+                    .iter()
+                    .map(|(k, v)| {
+                        Value::Structure("=".into(), vec![Value::Var(k.clone()), v.clone()])
+                    })
+                    .collect::<Vec<_>>(),
+            ),
+        }
+    }
+}
+
+impl From<&QueryResolution> for Value {
+    fn from(qr: &QueryResolution) -> Self {
+        match qr {
+            QueryResolution::True => Value::Atom("true".into()),
+            QueryResolution::False => Value::Atom("false".into()),
+            QueryResolution::Matches(matches) => Value::disjunction(
+                &matches
+                    .iter()
+                    .map(|m| {
+                        Value::conjunction(
+                            &m.bindings
+                                .iter()
+                                .map(|(k, v)| {
+                                    Value::Structure(
+                                        "=".into(),
+                                        vec![Value::Var(k.clone()), v.clone()],
+                                    )
+                                })
+                                .collect::<Vec<_>>(),
+                        )
+                    })
+                    .collect::<Vec<_>>(),
+            ),
+        }
+    }
+}
+
+struct ValueRationalInner {
+    up: IBig,
+    down: UBig,
```

Comment:
Maybe name these `nominator` and `denominator` instead of `up` and `down`

Reply:
Definitely! I named them that because I can never remember which is which.